Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,1422 @@
+        configuration: configuration,
+        customMetadata: metadata
+      )
+    case .clientClosedServerIdle(let state):
+      self.state = .clientClosedServerOpen(.init(previousState: state))
+      return self.makeResponseHeaders(
+        outboundEncoding: state.outboundCompression,
+        configuration: configuration,
+        customMetadata: metadata
+      )
+    case .clientIdleServerIdle:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+        "Client cannot be idle if server is sending initial metadata: it must have opened."
+      )
+    case .clientOpenServerClosed, .clientClosedServerClosed:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+        "Server cannot send metadata if closed."
+      )
+    case .clientOpenServerOpen, .clientClosedServerOpen:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+        "Server has already sent initial metadata."
+      )
+    }
+  }
+
+  private mutating func serverSend(message: [UInt8]) throws {
+    switch self.state {
+    case .clientIdleServerIdle, .clientOpenServerIdle, .clientClosedServerIdle:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+        "Server must have sent initial metadata before sending a message."
+      )
+    case .clientOpenServerOpen(var state):
+      state.framer.append(message)
+      self.state = .clientOpenServerOpen(state)
+    case .clientClosedServerOpen(var state):
+      state.framer.append(message)
+      self.state = .clientClosedServerOpen(state)
+    case .clientOpenServerClosed, .clientClosedServerClosed:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+        "Server can't send a message if it's closed."
+      )
+    }
+  }
+
+  private func makeTrailers(
+    status: Status?,
+    customMetadata: Metadata?,
+    trailersOnly: Bool
+  ) -> HPACKHeaders {
+    // If status isn't present, it means we're returning a non-200 HTTP :status
+    // header. This should already be included in the custom metadata: assert
+    // this and simply return those headers.
+    guard let status else {
+      assert(customMetadata != nil, "Something is very wrong")
+      var headers = HPACKHeaders()
+      for metadataPair in customMetadata! {
+        headers.add(name: metadataPair.key, value: metadataPair.value.encoded())
+      }
+      assert(headers.status != nil && headers.status != "200")
+      return headers
+    }
+
+    // Trailers always contain the grpc-status header, and optionally,
+    // grpc-status-message, and custom metadata.
+    // If it's a trailers-only response, they will also contain :status and
+    // content-type.
+    var headers = HPACKHeaders()
+    let customMetadataCount = customMetadata?.count ?? 0
+    if trailersOnly {
+      // Reserve 4 for capacity: 3 for the required headers, and 1 for the
+      // optional status message.
+      headers.reserveCapacity(4 + customMetadataCount)
```

Comment:
The conditional reserving seems a bit too heavily optimised in my opinion. I personally would just always reserve 4.